### DISCUSSION TITLE
Feature request: Auto-generation of relationship and support for --generate for relationships

### DISCUSSION THREAD

**Discussion by damms005** [filamentphp/filament#2428] [Ideas]
First, I'd love to mention that I am willing to work on this feature if the authors/contributors/community welcome my arguments below:

Thank you for this wonderful package. It is unmatched in Laravel history: Nova-grade and more, but for free.

The feature: we already have a `php artisan make:filament-resource <resource> --generate` command. I am of the opinion that we can add a hypothetical `--with-relations` option to the command.

This proposed `--with-relations` flag will simply intelligently run a corresponding `php artisan make:filament-<relationship-name> <resource> <relationship> <attribute> ` after generating the resource.

We may even start with a bare minimum for this functionality, meaning we do not need to support auto-generation of all Laravel relationships in the first release of this feature. We may support the most popular ones (opinionated): `HasOne`, `HasMany`, `BelongsTo`, and `BelongsToMany` (this may be highlighted in the documentation of this new feature).

I believe that this can be achieved by a simple [Reflection](https://www.php.net/manual/en/book.reflection.php) and an opinionated approach. This is my thought process on implementing a hypothetical auto-generation of relationships for a `Customer` that `BelongsTo` a `User` and `HasMany` `Orders`: 

- User runs `php artisan make:filament-resource Customer --generate --with-relations`
- [The handler](https://github.com/laravel-filament/filament/blob/9b1656e700d574b909ec2f0aca2f34cf19d6556c/packages/admin/src/Commands/MakeResourceCommand.php#L18) is called as usual, [but the relations logic](https://github.com/laravel-filament/filament/blob/9b1656e700d574b909ec2f0aca2f34cf19d6556c/packages/admin/src/Commands/MakeResourceCommand.php#L87) is replaced with a new logic that first checks for the `--with-relations` flag and abstracts the code into a new `$this->generateRelationships()` function.
- `generateRelationships()` will use [Reflection](https://www.php.net/manual/en/book.reflection.php) to get the functions defined in `App\Models\Customer`. 
- For each function, we can check if it is a supported relationship like the below:
- > `(new ReflectionClass('App\Models\Customer'))->getMethods()[function_index]->getReturnType()->getName()` (returns a FQCN)
- We can run a corresponding `php artisan make:filament-<relationship-name>` based on the relationship we get. The arguments to the relationship creation command can also be inferred using reflection. 
- For example, `(new ReflectionClass('App\Models\Customer'))->getMethods()[function_index]->getName()` will be used to get the relationship name.
- For the name of the attribute that will be used to identify the relationship, we can simply use the function name to infer the database table, get the column names, and intelligently pick one of the columns (or silently fail safely)
- We call the corresponding relationship generation  `php artisan make:filament-has-many CustomerResource <function_name> <inferred column>`.
- Finally, we can also have a `--generate` flag for relationships, such that instead of having a skeletal `form()` and `table()` for the generated relationships, we can look into how we can use `getResourceFormSchema()` and `getResourceTableColumns()` from `CanGenerateResources` trait to give it more life beyond the current skeletal default

Thoughts?

**Comment by danharrin** [comment]
Yes, that would be great! We could use an [option array](https://laravel.com/docs/9.x/artisan#option-arrays) to allow the user to define which relationships to generate -

```bash
make:filament-resource ... --relationship=posts --relationship--categories
```

`CanGenerateResources` was written very quickly and I did intend to expand it much more, so any refactoring you see is appropriate would be welcome.
